### ISSUE TITLE
High availability support for  Federation

### ISSUE THREAD

**Issue by shaneu** [apollographql/apollo-server#3540]
First let me say we love federation on our team. We split our monolith into approx. 11 microservices and have been very pleased with the ease of the refactor and the results.

There were some issues around availability that we had to roll ourselves, and it led me to think maybe these would be better included in the federation/gateway package. 

First of all, the gateway attempts to connect to the federated services, and if it is unable it still get's initialized, only without the services it failed to connect too. This is troublesome for an application that is deployed automatically as it requires human intervention to monitor if the gateway initialization was successful. 

Our workaround was to create a `readinessProbe` that performs the same query the gateway makes to the federated services, with the goal being to make sure the services are available and ready to receive traffic. This probe has a configurable number of retries before throwing. If the probe is unsuccessful we let the application to crash, as it doesn't make sense for us to allow it to initialize without all services being available. 

Perhaps this feature could be built into the gateway, along with a configurable number of retries? If it is unsuccessful it crashes and does not init. Perhaps that too could be a configuration `mustConnect: true/false` (I'm sure there's a better name for the property).

Other than that we let kubernetes manage the availability of the other services via liveness probes, but for teams not using a deployment orchestration manager like a kubernetes perhaps there is a way by which the gateway could check the availability of its downstream services?

Also, a periodic reloading of the schema would be helpful. Let's take the case where we update a schema in one of our services and redeploy it. The gateway wouldn't know the service's schema had changed so we must restart the gateway for it to get the changes. This hurts availably by causing downtime while the gateway is redeployed. If the gateway would refresh its schema on a configurable polling interval that issue could be avoided. We've developed a hack using helm charts to tell the gateway to redeploy when one of the downstream services have been updated, but that still causes downtime. 

If any of these ideas seems like it would be a good addition to the project I would be more than happy to make a PR. I am truly appreciate of the work the apollo team is doing and I'd love to contribute.

**Comment by moimikey** [comment]
upvoteeeee 👏 👏 👏 👏 👏

**Comment by rtymchyk** [comment]
It sounds to me like managed federation solves both of these issues? In managed mode, you will rarely need to re-deploy the gateway. Once it's up and running, it will get updated as federated services notify it via `service:push`.

**Comment by shaneu** [comment]
managed is an option for every team, mine included. It would be nice to have this built into the gateway or a service that can be deployed along side the gateway

**Comment by rtymchyk** [comment]
@shaneu can you explain why managed isn't an option for your team? I am just curious and playing the devil's advocate :)

**Comment by shaneu** [comment]
@rtymchyk I work for an enterprise bank that due to internal policy and compliance rules must manage all infrastructure on prem. Unless I'm incorrect, Apollo Graph Manager does not have an on prem solution. I'm sure my teams situation in those regards is not unique.

**Comment by alanhoff** [comment]
I'm also blocked by this... If a federated container needs to restart then the gateway becomes pretty much useless since it wont try to reconnect/reload and will just start throwing `503`

**Comment by rtymchyk** [comment]
@alanhoff Is this an an issue with federation? If your container needs to restart, you should be doing rolling restarts (or rolling updates for deployments). i.e. there shouldn't be a state where your service is down, regardless of whether or not it is used for federation.

**Comment by Quantumplation** [comment]
@rtymchyk 
That works in a perfect, but making rolling deployments a requirement of the infrastructure / CICD pipelines limits the audience significantly, and can be quite fragile.

When spinning up new ephemeral environments, for example, Apollo federation induces a startup order on services.  In very large organizations with hundreds of microservices, this kind of hard coded ordering of services is a PAIN to maintain.

As another example, when doing continuous deployment, a new version of a federated service means that you need to remember to redeploy the Apollo gateway(s) that depend on that service, hard coding this interdependency in the build pipeline, rather than in the service code itself.

Like @shaneu mentioned above, a managed-off-prem dependency is a non-starter for us.

Ideally, Apollo federation would have a mechanism to indicate "hey, the schema graph you constructed on startup may be out of date, please re-fetch from the same services.", to allow us to orchestrate the refresh policy in a way that meets our own needs.

**Comment by Glen-Moonpig** [comment]
Managed federation still doesn't re-initialiaze if it fails to pick up the schema. The polling timer never gets started so you just have a dud running service.

**Comment by AndreBtt** [comment]
@shaneu You said that you are manage the availability of your services using kubernetes, how are you updating the gateway when some service goes down or a new service is available ?

I'm asking you that because I'm also using kubernetes, and to manage these services I store their name and host in an array that I used to create the Apollo Gateway, like this:
`const gateway = new ApolloGateway({
  serviceList: services,
});`

Once a service goes down or goes up, I update this array and use `gateway.load();` to update gateway without turn the server down. I'm not comfortable with this solution because is seems that `load` function restart all my services not only the one changed.

Is there a better way to do this?
